lib/pavilion/builder.py

```diff
@@ -586,12 +592,20 @@ def _build(self, build_dir, cancel_event):
                                         stderr=build_log)
 
                 result = None
+                start = time.time()
+                timeout = self._timeout
                 while result is None:
                     try:
                         result = proc.wait(timeout=1)
                     except subprocess.TimeoutExpired:
-                        log_stat = self.tmp_log_path.stat()
-                        timeout = log_stat.st_mtime + self._timeout
+                        if self._timeout_file.exists():
+                            log_stat = self._timeout_file.stat()
+                            timeout = max(timeout, log_stat.st_mtime +
+                                          self._timeout)
+                        else:
+                            log_stat = self.tmp_log_path.stat()
+                            timeout = max(timeout, log_stat.st_mtime +
+                                          self._timeout)
```

Comment:
There's a race condition here that I just noticed. The file could disappear between checking for its existence and checking 'stat'. It's not likely, but if it happens things will explode.

```python
timeout_file = self.timeout_file if self.timeout_file.exists else self.tmp_log_path
try:
    timeout = max(timeout_file.stat().st_mtime + self._timeout,
                            timeout)
except OSError:
    pass
```